### PR TITLE
Add support for multi-select Drop downs

### LINKED ISSUE #1543: Multiple Value select for dropDown API
This is a feature request for selecting multiple values with the dropDown API

### Expected behavior
dropDown(label).select(option1,option2') ==> Should select multiple values
dropDown(label).select({index:0,1,2,3})
dropDown(label).select(/subString/) ==> Should select all the options which match the given Regex

### Actual behavior
dropDown(label).select(option1,option2') ==> throws error
dropDown(label).select({index:0,1,2,3}) ==> this selects only the last option ie index 3
dropDown(label).select(/subString/) ==> selects first matching regex

### PR DESCRIPTION
**Pull request by luthrap**
Fixes #1543 |Paras+Jyoti|

### CODE REVIEW

File: lib/elementWrapper/dropDownWrapper.js
Comment (zabil) [inline lib/elementWrapper/dropDownWrapper.js:55]:
```suggestion
   * @param {string[]} values value(s) to be selected
```

File: lib/elementWrapper/dropDownWrapper.js
Comment (zabil) [inline lib/elementWrapper/dropDownWrapper.js:54]:
```suggestion
   * Select option or options from a drop down
```

**Review by zabil — CHANGES_REQUESTED** [review]
This is not a full review.

Notice that the docs can be modified to reflect the right parameter type.

**Comment by luthrap** [comment]
Will do that. Also, had to add some extra conditions to support the test case in unit-tests/elements/dropDown.test.js. The conditions were added as we have changed the CDP command to select a drop down value to support multi-select which was not getting supported in the current node structure.

File: lib/elements/dropDown.js
Comment (saikrishna321) [inline lib/elements/dropDown.js:111]: Dead code. Can this be removed?

File: lib/elements/dropDown.js
Comment (saikrishna321) [inline lib/elements/dropDown.js:115]: What is the need to set null and undefined?

File: lib/elements/dropDown.js
Comment (saikrishna321) [inline lib/elements/dropDown.js:194]: Why can't we just return the array `selectedValues `?

**Review by saikrishna321 — CHANGES_REQUESTED** [review]
Few changes needed!

File: lib/elements/dropDown.js
Comment (luthrap) [reply lib/elements/dropDown.js:111]: Although it was written to skip other elseif conditions if the value provided is an array, will refactor it.

File: lib/elements/dropDown.js
Comment (luthrap) [reply lib/elements/dropDown.js:194]: If we just return array, it will not be backward compatible for which single value was provided.

Comment (saikrishna321) [reply lib/elements/dropDown.js:190]: Thought - Should we have a new function called `values()` which would return multiple values in array format. 
@NivedhaSenthil @zabil

File: lib/elementWrapper/dropDownWrapper.js
Comment (zabil) [inline lib/elementWrapper/dropDownWrapper.js:55]: Can you also add a few examples usages of using array as parameters? You can add it here using the `@example` tag and at `dropDown` API jsdoc which will then appear at https://docs.taiko.dev/api/dropdown/

File: lib/elements/dropDown.js
Comment (luthrap) [reply lib/elements/dropDown.js]: Fixed. Please check.

File: lib/elementWrapper/dropDownWrapper.js
Comment (luthrap) [reply lib/elementWrapper/dropDownWrapper.js:55]: Added.

File: lib/elements/dropDown.js
Comment (luthrap) [reply lib/elements/dropDown.js]: Fixed.

File: lib/elements/dropDown.js
Comment (zabil) [reply lib/elements/dropDown.js:190]: > Thought - Should we have a new function called values()

You mean selected values right? Yeah, that makes sense. Can be part of another PR. @saikrishna321 can you create an issue for that?

Comment (saikrishna321) [reply lib/elements/dropDown.js:190]: @zabil Yes, for selected values. Will create a new issue for that.

Comment (saikrishna321) [reply lib/elements/dropDown.js:190]: @luthrap We can address this here  https://github.com/getgauge/taiko/issues/1613. Can this be reverted?

Comment (luthrap) [reply lib/elements/dropDown.js:190]: Sure, will revert and update the test cases. Meanwhile, it appears that a lot of extra files have been checked in as part of the PR. This happened while signing off the previous commits, I mistakenly signed in those commits that should not belong to this PR. Could it be removed from here or shall i open a new clean PR request.

Comment (zabil) [reply lib/elements/dropDown.js:190]: > Sure, will revert and update the test cases.

Sorry, if it's been already implemented let's keep it! Maybe refer to the issue in this PR

> Could it be removed from here or shall i open a new clean PR request.

I noticed that too. You could update the same PR by fixing the merge and arranging the commits.

Comment (saikrishna321) [reply lib/elements/dropDown.js:190]: >if it's been already implemented let's keep it

@zabil It's done as a part of current function `value()`

Comment (luthrap) [reply lib/elements/dropDown.js:190]: We can move the value() code of the PR request in the values() function and have the original logic in the value() function. This will require a change in a couple of test cases and would conclude the PR correctly.

Comment (zabil) [reply lib/elements/dropDown.js:190]: > It's done as a part of current function value()

IMHO that should be fine. It might be confusing to have two different API's to return values depending on the type of the box when there is a common API to set the values depending on the type of the combo box.

So I am assuming `value` returns and array when there are multiple values selected

@saikrishna321 luthrap What do you think?

Comment (luthrap) [reply lib/elements/dropDown.js:190]: Agreed, we should be encapsulating the logic in one function only.
```
if(selectedValues.length <= 1){
  return selectedValues.join()
else{
  return selectedValues
}
```

Comment (saikrishna321) [reply lib/elements/dropDown.js:190]: > So I am assuming value returns an array when there are multiple values selected

Agree!

File: lib/elements/dropDown.js
Comment (saikrishna321) [inline lib/elements/dropDown.js:152]: Can you add a test to validate this?

**Review by saikrishna321 — CHANGES_REQUESTED** [review]

File: lib/elements/dropDown.js
Comment (luthrap) [reply lib/elements/dropDown.js:152]: The tests has already been already in dropDown.test.js. For the disabled field, following test are added :
```
 it('test select() should throw when selecting multiple values and one of the value is disabled', async () => {
      try {
        expect(await dropDown('Foods').select(['Burger', 'Pasta']));
      } catch (err) {
        expect(err.message).to.equal('Cannot set value Burger on a disabled field');
        expect(await dropDown('Foods').value()).to.equal('');
      }
    });
```

Test for undefined was already present earlier in the master branch 
```
it('should throw error for undefined values', async () => {
      try {
        expect(await dropDown('Cars').select(undefined));
      } catch (err) {
        expect(err.message).to.equal('Option undefined not available in drop down');
      }
    });
```

**Review by zabil — APPROVED** [review]

**Comment by zabil** [comment]
@saikrishna321 will wait for your approval.

The PR looks good to me. Great work! @luthrap @guptajyoti845

**Comment by luthrap** [comment]
@zabil : I am yet to commit the changes for `value()` function. Will do it by EOD. If @saikrishna321 has a go ahead, PR can be merged once i update the `value()` function.

**Comment by saikrishna321** [comment]
@luthrap @guptajyoti845 Thanks a lot for working on all the reviews. 

@zabil LGTM! @luthrap Once the `value()` is fixed. Happy to merge

**Comment by luthrap** [comment]
Pushed the changes.

**Update** 
If we look at the signature below, it returns a `Promise<string>`. However, we should return `Promise<string | Array<string> | Array<number>>` as this is the correct way. Should we add this? Although adding this will start breaking other tests from `htmlElementAPI.ts`
```
export interface ValueWrapper extends ElementWrapper {
  value(): Promise<string>;
}
```

**Comment by zabil** [comment]
> However, we should return `Promise<string | Array<string> | Array<number>>` as this is the correct way although it does not impact the functionality. Should we add this? Adding this will start breaking other tests from `htmlElementAPI.ts`.

Adding @fmpanelli to the discussion.

**Review by saikrishna321 — APPROVED** [review]

**Comment by luthrap** [comment]
@zabil @saikrishna321 : can it be merged and the issue related to the change in contract can be picked up in other issue.

**Comment by zabil** [comment]
Agree. I'll test this a bit more today and merge it.

**Comment by fmpanelli** [comment]
I am having a look at the types...

**Pull request merged**
